# Group 2 Final Project - NYC Motor Vehicle Collisions Crashes

### Ask1 - Search for a dataset

**Identify and describe our dataset** 

- For our project, we intend to analyze New York Motor Vehicle Collisions. The dataset provides comprehensive details based on police reports for motor vehicle collisions in NYC.
- Each row in the dataset represents a crash event, including information such as the crash date, time, location, contributing factors, vehicle types, and outcomes.
- Dataset Size: 436.6 MB, Last Update: 11/28/2023, 2,045,723 records at the transactional level, and 29 variables.


**Dataset Source**

The Home of the U.S. Government's Open Database, New York City Police Department (NYPD). https://catalog.data.gov/dataset/motor-vehicle-collisions-crashes

**Why is important and what appeals to you about this dataset?**

The chosen dataset holds substantial potential for yielding valuable analytical insights. By delving into the patterns of motor vehicle collisions, we anticipate gaining a comprehensive understanding of the contributing factors and outcomes associated with these incidents.
      
Analyzing temporal patterns, spatial distributions, and identifying contributing factors such as vehicle types can uncover trends and correlations that contribute to enhancing traffic safety. Insights derived from this dataset can inform evidence-based decision-making, enabling us to propose targeted interventions and preventive measures to mitigate the occurrence and severity of motor vehicle collisions. This aligns with the broader goal of leveraging analytics for actionable insights.

**Acquire data**

In [1]:
!pwd

/home/ubuntu/notebooks/FinalProjectGroup2


In [2]:
#The wget command to acquire the dataset
!wget 'https://data.cityofnewyork.us/api/views/h9gi-nx95/rows.csv?accessType=DOWNLOAD' -O Motor_Vehicle_Collisions_-_Crashes.csv 
#Count the number of ,lines (rows) in the raw data; the result should be 2045723.
!wc -l Motor_Vehicle_Collisions_-_Crashes.csv

2045723 Motor_Vehicle_Collisions_-_Crashes.csv


In [19]:
# Rename the file. 
!mv Motor_Vehicle_Collisions_-_Crashes.csv MVC.csv

In [12]:
#Check data
!csvclean -n MVC.csv

No errors.


In [2]:
#Check the column names
!csvcut -n MVC.csv

  1: CRASH DATE
  2: CRASH TIME
  3: BOROUGH
  4: ZIP CODE
  5: LATITUDE
  6: LONGITUDE
  7: LOCATION
  8: ON STREET NAME
  9: CROSS STREET NAME
 10: OFF STREET NAME
 11: NUMBER OF PERSONS INJURED
 12: NUMBER OF PERSONS KILLED
 13: NUMBER OF PEDESTRIANS INJURED
 14: NUMBER OF PEDESTRIANS KILLED
 15: NUMBER OF CYCLIST INJURED
 16: NUMBER OF CYCLIST KILLED
 17: NUMBER OF MOTORIST INJURED
 18: NUMBER OF MOTORIST KILLED
 19: CONTRIBUTING FACTOR VEHICLE 1
 20: CONTRIBUTING FACTOR VEHICLE 2
 21: CONTRIBUTING FACTOR VEHICLE 3
 22: CONTRIBUTING FACTOR VEHICLE 4
 23: CONTRIBUTING FACTOR VEHICLE 5
 24: COLLISION_ID
 25: VEHICLE TYPE CODE 1
 26: VEHICLE TYPE CODE 2
 27: VEHICLE TYPE CODE 3
 28: VEHICLE TYPE CODE 4
 29: VEHICLE TYPE CODE 5


In [83]:
!csvgrep -c1 -r "201[6789]|202[012]" MVC.csv > MVC_filtered.csv

**Key Decision we took to modify/remove data**

- While the original dataset we downloaded spans a broad timeframe, from 1998 to 2022, the source website (https://catalog.data.gov/dataset/motor-vehicle-collisions-crashes) indicates a significant change in data collection methods, transitioning from manual entry to electronic entry. Therefore, to address data entry inconsistencies, we have opted to narrow our analysis to the years 2016 to 2020. This allows us to focus on a specific period, providing a more concentrated and relevant dataset for our analysis.

In [2]:
#Count the number of lines (rows) in the filtered file; the result should be 1231230.
!wc -l MVC_filtered.csv

1231230 MVC_filtered.csv


**Perform initial exploration**

Using 'csvstat' to generate statistical information, this includes variable data types, and it helps identify columns that may contain null values.
**Note: This code should take at least 10min to run.** 

In [46]:
!csvstat MVC_filtered.csv

/home/ubuntu/.local/lib/python3.8/site-packages/agate/table/from_csv.py:74: RuntimeWarning: Error sniffing CSV dialect: Could not determine delimiter
  1. "CRASH DATE"

	Type of data:          Date
	Contains null values:  False
	Unique values:         2557
	Smallest value:        2016-01-01
	Largest value:         2022-12-31
	Most common values:    2018-11-15 (1065x)
	                       2017-12-15 (999x)
	                       2017-05-19 (974x)
	                       2017-05-18 (911x)
	                       2017-01-07 (896x)

  2. "CRASH TIME"

	Type of data:          TimeDelta
	Contains null values:  False
	Unique values:         1440
	Smallest value:        0:00:00
	Largest value:         0:23:59
	Sum:                   11548 days, 9:08:55
	Most common values:    0:00:00 (18717x)
	                       0:16:00 (17339x)
	                       0:17:00 (16956x)
	                       0:15:00 (16631x)
	                       0:14:00 (15825x)

  3. "BOROUGH"

	Type of data:     

After we acquire data and perform initial exploration by using "csvstat" command.
Based on the provided statistical information, we are reasonably sure the dataset appears suitable for dimensional modeling and analytical analysis for the following reasons:

**"CRASH DATE" and "CRASH TIME":**
"CRASH DATE" contains date values ranging from 2016-01-01 to 2022-12-31, and "CRASH TIME" contains time values. Both columns have no null values and provide a detailed temporal aspect for analysis.

**"BOROUGH" and "ZIP CODE":**
These columns contain categorical and numerical location information, respectively. While "BOROUGH" has null values, they are excluded from calculations. Both columns offer geographic context for analysis.

**"LATITUDE" and "LONGITUDE":**
These numerical columns provide precise location coordinates. Although they have null values, they are excluded from calculations. They can be used for spatial analysis.

**"NUMBER OF PERSONS INJURED" and "NUMBER OF PERSONS KILLED":**
These numerical columns contain information on injuries and fatalities in accidents, allowing for safety analysis.

**"CONTRIBUTING FACTOR VEHICLE 1" to "CONTRIBUTING FACTOR VEHICLE 5":**
These categorical columns provide information on contributing factors in accidents, allowing for analysis of the causes of incidents.

**"VEHICLE TYPE CODE 1" to "VEHICLE TYPE CODE 5":**
These categorical columns provide information on the types of vehicles involved in accidents, enabling analysis based on vehicle types.

**"COLLISION_ID":**
This numerical column serves as a unique identifier for each collision, allowing for easy referencing in analysis.

Overall, the dataset offers a diverse set of features, both numerical and categorical, with a substantial number of entries (1231230 rows), making it suitable for comprehensive dimensional modeling and analytical exploration.

**Analytical questions we can answer with this data**

We are a consultation firm providing an overview of the New York City traffic incident trends from 2016-2022. Our goal is to offer suggestions to car insurance companies, the New York City Police (NYPD) Traffic Enforcement, and the New York Department of Motor Vehicles (DMV).

Analysis 1 for Car Insurance:
- Which top 10 vehicle types have the most crashes?
- In which boroughs do the most crashes occur?

Analysis 2 for NYPD Traffic Enforcement:
- Which top 10 zip code areas have the highest number of persons injured due to crashes?
- What combination of day of the week, hour of the day, and zip code experienced the top 30 highest numbers of persons injured in crashes?

Analysis 3 for New York Department of Motor Vehicles (DMV):
- Identify the top 10 contributing factors to the highest number of fatalities.

**Concerns with the data and changes we expect to overcome**

**Concerns**
- Data Completeness: There might be potential missing values, particularly in crucial fields such as boroughs, zip codes, contributing factors, and vehicle types.
- Data Volume/Computational Issues: Extracting data from 2016-2022 results in around 1,200,000 records, potentially leading to slow processing times for codes.

**Expected Changes:** 
We anticipate addressing these concerns by dropping null values in critical fields and eliminating unnecessary columns.


### Ask2 - Data Wrangling & Data Cleaning 

The steps for Ask 2 are as follows:

1. Creation of Fact Table:

    We initiate the process by creating a fact table named "Crashes" and utilize the COPY command to import all data into our primary fact table.

2. Column Selection and Cleanup:

    Next, we streamline our dataset by eliminating unnecessary columns in our data analysis, such as latitude, longitude, location, on_street_name, cross_street_name, off_street_name, number_of_pedestrians_injured, number_of_pedestrians_killed, number_of_cyclist_injured, number_of_cyclist_killed, number_of_motorist_injured, and number_of_motorist_killed.

3. Handling Null Values:

    We proceed by removing null values in the 'borough,' 'zip_code,' and 'contributing_factor_vehicle_1' columns. Given that these columns are crucial for our analytics, any rows with null values in these columns are dropped.

4. Star Schema Design:

    Our modeling approach involves creating a star schema. We identify the fact table as "Crashes" and establish four dimensional tables: Geography, Contributing factor, Vehicle type, and Crash time.

5. Dimensional Table Processing:

    We delve into working on the four dimensional tables (Geography, Contributing factor, Vehicle type, and Crash time). Simultaneously, we modify the fact table and establish links between them.

6. Final Fact Table:
    Upon completing the above steps, we arrive at our final fact table, which integrates the cleaned and processed data from both the fact and dimensional tables.

In [4]:
%load_ext sql

In [26]:
!dropdb -U student Final_Project

In [27]:
!createdb -U student Final_Project

In [5]:
%sql postgresql://student@/Final_Project

The above code cells %load_text sql allows for the %%sql magic command to used through out ther notebook. To have the next tasks completed you will need to run this command line. The following codes asks that you drop and create the database Final_Project. These two command will not need to be executed once they have been executed before. The final code line "%sql postgresql://student@/Final_Project creates the link between the database you've created and the notebook you are currently working in. 

In [6]:
%%sql

DROP TABLE IF EXISTS Crashes;

CREATE TABLE Crashes (    
    CRASH_DATE DATE NOT NULL,
    CRASH_TIME TIME(20) NOT NULL,
    BOROUGH CHAR(20),
    ZIP_CODE NUMERIC(10),
    LATITUDE NUMERIC(20),
    LONGITUDE NUMERIC(20),
    LOCATION CHAR(30),
    ON_STREET_NAME CHAR(50),
    CROSS_STREET_NAME CHAR(50),
    OFF_STREET_NAME CHAR(50),
    NUMBER_OF_PERSONS_INJURED NUMERIC(10),
    NUMBER_OF_PERSONS_KILLED NUMERIC(10),
    NUMBER_OF_PEDESTRIANS_INJURED NUMERIC(10),
    NUMBER_OF_PEDESTRIANS_KILLED NUMERIC(10),
    NUMBER_OF_CYCLIST_INJURED NUMERIC(10),
    NUMBER_OF_CYCLIST_KILLED NUMERIC(10),
    NUMBER_OF_MOTORIST_INJURED NUMERIC(10),
    NUMBER_OF_MOTORIST_KILLED NUMERIC(10),
    CONTRIBUTING_FACTOR_VEHICLE_1 CHAR(100),
    CONTRIBUTING_FACTOR_VEHICLE_2 CHAR(100),
    CONTRIBUTING_FACTOR_VEHICLE_3 CHAR(100),
    CONTRIBUTING_FACTOR_VEHICLE_4 CHAR(100),
    CONTRIBUTING_FACTOR_VEHICLE_5 CHAR(100),
    COLLISION_ID NUMERIC(10) NOT NULL,
    VEHICLE_TYPE_CODE_1 CHAR(100),
    VEHICLE_TYPE_CODE_2 CHAR(100),
    VEHICLE_TYPE_CODE_3 CHAR(100),
    VEHICLE_TYPE_CODE_4 CHAR(100),
    VEHICLE_TYPE_CODE_5 CHAR(100)
)

 * postgresql://student@/Final_Project
Done.
Done.


[]

The simplest way to insert the data into one table instead of each dimensional table. This also allowed for us to easily clean the data, by having one base table then start removing null values shown later in the notebook. 

In [4]:
%%sql
SELECT * from Crashes
Limit 10

 * postgresql://student@/Final_Project
10 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               ', 167, 38, 1108, 125468),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               ', 217, 38, 1108, 102912),
 (datetime.date(2021, 12, 14), datetime.time(12, 54), 'BROOKLYN            ', Decimal('11217'), Decimal('1'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4487052'), 'Sedan                                                                                               ', 174, 38, 1108, 204230),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               ', 8, 16, 1108, 260466),
 (datetime.date(2021, 12, 14), datetime.time(16, 50), 'QUEENS              ', Decimal('11413'), Decimal('0'), Decimal('0'), 'Turning Improperly                                                                                  ', Decimal('4487127'), 'Sedan                                                                                               ', 69, 34, 1108, 320759),
 (datetime.date(2021, 12, 14), datetime.time(17, 15), 'BROOKLYN            ', Decimal('11211'), Decimal('1'), Decimal('0'), 'Passing or Lane Usage Improper                                                                      ', Decimal('4486556'), 'Bus                                                                                                 ', 145, 21, 1007, 352811),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               ', 201, 15, 1108, 456487),
 (datetime.date(2021, 12, 14), datetime.time(23, 10), 'QUEENS              ', Decimal('11434'), Decimal('2'), Decimal('0'), 'Reaction to Uninvolved Vehicle                                                                      ', Decimal('4486635'), 'Sedan                                                                                               ', 59, 56, 1108, 493185),
 (datetime.date(2022, 3, 21), datetime.time(12, 5), 'MANHATTAN           ', Decimal('10018'), Decimal('1'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4514237'), 'Motorcycle                                                                                          ', 120, 32, 582, 210587),
 (datetime.date(2022, 3, 26), datetime.time(5, 13), 'QUEENS              ', Decimal('11418'), Decimal('0'), Decimal('0'), 'Failure to Yield Right-of-Way                                                                       ', Decimal('4513470'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 139, 61, 1140, 59664)]

The table above shows the top 10 rows of the dataset from the table CRASHES. This helped us to understand which columns and rows were sum columns, such as all the injured and fatalities columns were summed in the total injured and total fatalities, and that of the contributing factors, and vehicle types. (The limit of 10 was to shorten the notebook for easier reading, but can be changed to the entire dataset of 793560, as shown below).

In [ ]:
%%sql
SELECT * from Crashes

 * postgresql://student@/Final_Project
793560 rows affected.


As seen the total number of rows have now lessened to 793560 from 1231230. This strictly accounts for all the rows and columns related to the Crashes table.

In [7]:
!pwd

/home/ubuntu/notebooks/FinalProjectGroup2


**Importing Bulk Data**

In [8]:
%%sql
COPY Crashes FROM '/home/ubuntu/notebooks/FinalProjectGroup2/MVC_filtered.csv'
CSV
HEADER;

 * postgresql://student@/Final_Project
1231229 rows affected.


[]

This code is vital to the dimensional tables, with this command the dataset MVC_filtered.csv is inserted into our base table "CRASHES". Without inserting data into the table, the codes and commands entered will read header or column headers, but will not read the data in the rows preceeding. The commands and codes truly run successfully only after inserting the data.  

In [ ]:
%%sql
SELECT * from Crashes
LIMIT 5

 * postgresql://student@/Final_Project
793560 rows affected.


Drop unnecessary columns

In [9]:
%%sql
ALTER TABLE Crashes
DROP COLUMN latitude,
DROP COLUMN longitude,
DROP COLUMN location,
DROP COLUMN on_street_name,
DROP COLUMN cross_street_name,
DROP COLUMN off_street_name;

 * postgresql://student@/Final_Project
Done.


[]

In [10]:
%%sql
ALTER TABLE Crashes
DROP COLUMN number_of_pedestrians_injured,
DROP COLUMN number_of_pedestrians_killed,
DROP COLUMN number_of_cyclist_injured,
DROP COLUMN number_of_cyclist_killed, 
DROP COLUMN number_of_motorist_injured,
DROP COLUMN number_of_motorist_killed;

 * postgresql://student@/Final_Project
Done.


[]

In [11]:
%%sql
ALTER TABLE Crashes
DROP COLUMN contributing_factor_vehicle_2,
DROP COLUMN contributing_factor_vehicle_3,
DROP COLUMN contributing_factor_vehicle_4,
DROP COLUMN contributing_factor_vehicle_5, 
DROP COLUMN vehicle_type_code_2,
DROP COLUMN vehicle_type_code_3,
DROP COLUMN vehicle_type_code_4,
DROP COLUMN vehicle_type_code_5;

 * postgresql://student@/Final_Project
Done.


[]

The above three cells drop command tells the table to remove the following columns: latitude, longitude, location, on_street_name, cross_street_name, off_street_name, number_of_pedestrians_injured, number_of_pedestrians_killed, number_of_cyclist_injured, number_of_cyclist_killed, number_of_motorist_injured, number_of_motorist_killed. However, we specifically did so for the following reasons. Regarding the location values is null and the zip code/ Borough provides us enough information geographically to set a dimensional table "Geography". The number of pedestrians, cyclist, motorist, killed and injured were consolidated into two columns total injured and total killed respectively. For contributing factor 2,3,4,and 5, were either duplicates of contributing factor 1 or held no additional information, the same can be said about the vehicle type.  

**The biggest challenge we encountered was updating the contributing_factor_key and vehicle_type_key in the Crashes table.**

- Reason 1:
When attempting to match all five values for contributing_factor_vehicle_1 and vehicle_type_1 simultaneously, the execution time was 30 minutes.

- Reason 2:
Contributing factors 2, 3, 4, and 5, as well as vehicle types, either duplicated contributing factor 1 or provided no additional information; the same can be said about the vehicle types.

Consequently, we decided to retain only the first contributing factor and vehicle type for matching in the table. Simultaneously, we shifted our focus to the vehicle causing the crash rather than the affected vehicle.

In [12]:
%%sql
SELECT * from Crashes
LIMIT 5

 * postgresql://student@/Final_Project
5 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(2, 39), None, None, Decimal('2'), Decimal('0'), 'Aggressive Driving/Road Rage                                                                        ', Decimal('4455765'), 'Sedan                                                                                               '),
 (datetime.date(2022, 3, 26), datetime.time(11, 45), None, None, Decimal('1'), Decimal('0'), 'Pavement Slippery                                                                                   ', Decimal('4513547'), 'Sedan                                                                                               '),
 (datetime.date(2022, 6, 29), datetime.time(6, 55), None, None, Decimal('0'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4541903'), 'Sedan                                                                                               '),
 (datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(8, 13), 'BROOKLYN            ', Decimal('11233'), Decimal('0'), Decimal('0'), None, Decimal('4486609'), None)]

In [16]:
%%sql
SELECT * from Crashes
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(2, 39), None, None, Decimal('2'), Decimal('0'), 'Aggressive Driving/Road Rage                                                                        ', Decimal('4455765'), 'Sedan                                                                                               '),
 (datetime.date(2022, 3, 26), datetime.time(11, 45), None, None, Decimal('1'), Decimal('0'), 'Pavement Slippery                                                                                   ', Decimal('4513547'), 'Sedan                                                                                               '),
 (datetime.date(2022, 6, 29), datetime.time(6, 55), None, None, Decimal('0'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4541903'), 'Sedan                                                                                               '),
 (datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(8, 13), 'BROOKLYN            ', Decimal('11233'), Decimal('0'), Decimal('0'), None, Decimal('4486609'), None),
 (datetime.date(2021, 4, 14), datetime.time(12, 47), None, None, Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4407458'), 'Dump                                                                                                '),
 (datetime.date(2021, 12, 14), datetime.time(17, 5), None, None, Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486555'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               ')]

Delete the null values in the 'borough,' 'zip_code,' and 'contributing_factor_vehicle_1' columns. Since these columns are our main focus for analytics, any rows with null values in these columns will be dropped.

In [21]:
%%sql
SELECT count(*)
From Crashes 
where borough is NULL;

 * postgresql://student@/Final_Project
1 rows affected.


[(0,)]

In [14]:
%%sql
DELETE FROM Crashes
WHERE borough IS NULL;

 * postgresql://student@/Final_Project
434027 rows affected.


[]

In [22]:
%%sql
SELECT count(*)
From Crashes 
where zip_code is NULL;

 * postgresql://student@/Final_Project
1 rows affected.


[(0,)]

In [16]:
%%sql
DELETE FROM Crashes
WHERE zip_code IS NULL;

 * postgresql://student@/Final_Project
214 rows affected.


[]

In [23]:
%%sql
SELECT count(*)
From Crashes 
where contributing_factor_vehicle_1 is NULL;

 * postgresql://student@/Final_Project
1 rows affected.


[(0,)]

In [18]:
%%sql
DELETE FROM Crashes
WHERE contributing_factor_vehicle_1 IS NULL;

 * postgresql://student@/Final_Project
3428 rows affected.


[]

The above cells is the cleaning of the data aspect. This allows the team to work with only relevant data that can improve the query and answer questions from insurance companies along with traffic agencies. 

Count the final number of data for analysis after dropping the null values.

In [19]:
%%sql
SELECT count(*)
From Crashes

 * postgresql://student@/Final_Project
1 rows affected.


[(793560,)]

In [20]:
%%sql
SELECT * from Crashes
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(16, 50), 'QUEENS              ', Decimal('11413'), Decimal('0'), Decimal('0'), 'Turning Improperly                                                                                  ', Decimal('4487127'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(23, 10), 'QUEENS              ', Decimal('11434'), Decimal('2'), Decimal('0'), 'Reaction to Uninvolved Vehicle                                                                      ', Decimal('4486635'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 14), datetime.time(17, 58), 'BROOKLYN            ', Decimal('11217'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486604'), 'Tanker                                                                                              '),
 (datetime.date(2021, 12, 14), datetime.time(20, 3), 'BROOKLYN            ', Decimal('11226'), Decimal('4'), Decimal('0'), 'Steering Failure                                                                                    ', Decimal('4486991'), 'Sedan                                                                                               '),
 (datetime.date(2021, 12, 11), datetime.time(19, 43), 'BRONX               ', Decimal('10463'), Decimal('1'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4487040'), 'Station Wagon/Sport Utility Vehicle                                                                 '),
 (datetime.date(2021, 12, 11), datetime.time(4, 45), 'MANHATTAN           ', Decimal('10001'), Decimal('0'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4486905'), 'Station Wagon/Sport Utility Vehicle                                                                 ')]

The above shows the top 10 rows of data which allows us to review if the nulls have been removed, and if we were to run a query what to expect as output.

In [24]:
!pwd

/home/ubuntu/notebooks/FinalProjectGroup2


**We start with the Schema from this raw data with all the columns**

In [25]:
from IPython.display import Image

image_url = 'https://i.ibb.co/z2SJWcJ/Whats-App-Image-2023-12-09-at-11-40-33-AM.jpg'

Image(url=image_url)

**We eventually build a star-schema like this:**

In [11]:
image_url = 'https://i.ibb.co/ZhkWn2k/Dimensions.jpg'
Image(url=image_url)

**Work on Geography dimension, modify fact table and link them together**

In [28]:
%%sql
DROP TABLE IF EXISTS Geography;

CREATE TABLE Geography ( 
    key SERIAL PRIMARY KEY, 
    borough CHAR(20),
    zip_code NUMERIC(10)
);

 * postgresql://student@/Final_Project
Done.
Done.


[]

The first step to creating a dimensional table, we must use the create table command and create the primary key with the columns with their data value/limit. 

In [29]:
%%sql
INSERT INTO Geography (borough,zip_code) 
SELECT DISTINCT borough,zip_code
FROM Crashes;

 * postgresql://student@/Final_Project
234 rows affected.


[]

Now to ensure that we have data in the table, we'll insert the borough and zip code columns from our base table "Crashes".

In [30]:
%%sql
SELECT * From Geography
Order By key
limit 10

 * postgresql://student@/Final_Project
10 rows affected.


[(1, 'BRONX               ', Decimal('10465')),
 (2, 'BROOKLYN            ', Decimal('11226')),
 (3, 'QUEENS              ', Decimal('11416')),
 (4, 'MANHATTAN           ', Decimal('10169')),
 (5, 'BRONX               ', Decimal('10464')),
 (6, 'STATEN ISLAND       ', Decimal('10305')),
 (7, 'MANHATTAN           ', Decimal('10032')),
 (8, 'MANHATTAN           ', Decimal('10017')),
 (9, 'BROOKLYN            ', Decimal('11215')),
 (10, 'MANHATTAN           ', Decimal('10281'))]

Just checking if the Geography table functions correctly.

In [31]:
%%sql
SELECT count(*) From Geography
Where borough = 'MANHATTAN';

 * postgresql://student@/Final_Project
1 rows affected.


[(88,)]

In [32]:
%%sql
SELECT count(*) From Geography
Where borough = 'QUEENS'

 * postgresql://student@/Final_Project
1 rows affected.


[(67,)]

In [33]:
%%sql
ALTER TABLE Crashes
ADD COLUMN geography_key INTEGER, 
ADD CONSTRAINT fk_geography
FOREIGN KEY (geography_key) REFERENCES Geography (key);

 * postgresql://student@/Final_Project
Done.


[]

After effectively verfying that the table has data, we can add the contraint key.

In [34]:
%%sql
SELECT * from Crashes
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               ', None),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               ', None),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               ', None),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               ', None),
 (datetime.date(2021, 12, 14), datetime.time(16, 50), 'QUEENS              ', Decimal('11413'), Decimal('0'), Decimal('0'), 'Turning Improperly                                                                                  ', Decimal('4487127'), 'Sedan                                                                                               ', None),
 (datetime.date(2021, 12, 14), datetime.time(23, 10), 'QUEENS              ', Decimal('11434'), Decimal('2'), Decimal('0'), 'Reaction to Uninvolved Vehicle                                                                      ', Decimal('4486635'), 'Sedan                                                                                               ', None),
 (datetime.date(2021, 12, 14), datetime.time(17, 58), 'BROOKLYN            ', Decimal('11217'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486604'), 'Tanker                                                                                              ', None),
 (datetime.date(2021, 12, 14), datetime.time(20, 3), 'BROOKLYN            ', Decimal('11226'), Decimal('4'), Decimal('0'), 'Steering Failure                                                                                    ', Decimal('4486991'), 'Sedan                                                                                               ', None),
 (datetime.date(2021, 12, 11), datetime.time(19, 43), 'BRONX               ', Decimal('10463'), Decimal('1'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4487040'), 'Station Wagon/Sport Utility Vehicle                                                                 ', None),
 (datetime.date(2021, 12, 11), datetime.time(4, 45), 'MANHATTAN           ', Decimal('10001'), Decimal('0'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4486905'), 'Station Wagon/Sport Utility Vehicle                                                                 ', None)]

In [35]:
%%sql
SELECT * from Geography
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(1, 'BRONX               ', Decimal('10465')),
 (2, 'BROOKLYN            ', Decimal('11226')),
 (3, 'QUEENS              ', Decimal('11416')),
 (4, 'MANHATTAN           ', Decimal('10169')),
 (5, 'BRONX               ', Decimal('10464')),
 (6, 'STATEN ISLAND       ', Decimal('10305')),
 (7, 'MANHATTAN           ', Decimal('10032')),
 (8, 'MANHATTAN           ', Decimal('10017')),
 (9, 'BROOKLYN            ', Decimal('11215')),
 (10, 'MANHATTAN           ', Decimal('10281'))]

In [36]:
%%sql
UPDATE Crashes
SET geography_key = Geography.key
FROM Geography
WHERE Crashes.borough = Geography.borough
    AND Crashes.zip_code = Geography.zip_code;

 * postgresql://student@/Final_Project
793560 rows affected.


[]

With the constraint key created we need to update the 'Crashes' table by having the 'Crashes' table columns: 'borough' and 'zip_code' match that of 'Geographys' columns: 'borough' and 'zip_code' values. 

In [37]:
%%sql
SELECT * FROM Crashes 
LIMIT 5;

 * postgresql://student@/Final_Project
5 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               ', 167),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               ', 217),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               ', 201),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               ', 8),
 (datetime.date(2021, 12, 14), datetime.time(16, 50), 'QUEENS              ', Decimal('11413'), Decimal('0'), Decimal('0'), 'Turning Improperly                                                                                  ', Decimal('4487127'), 'Sedan                                                                                               ', 69)]

Check there are no null values in the 'geography_key' column.

In [38]:
%%sql 
SELECT count(*) From Crashes
Where geography_key is NULL

 * postgresql://student@/Final_Project
1 rows affected.


[(0,)]

We check that there are no null values in the 'Crashes' table in the 'geography_key'.

In [39]:
%%sql
SELECT count(DISTINCT geography_key)
FROM Crashes

 * postgresql://student@/Final_Project
1 rows affected.


[(234,)]

In [40]:
%%sql
SELECT count(key)
FROM Geography

 * postgresql://student@/Final_Project
1 rows affected.


[(234,)]

Check the distinct 'geography_key' count in both the fact table (Crashes) and the dimensional table (Geography). They have the same count, which is 234. Thus, we ensure that the fact table contains the right number of records.

**Work on Contributing_factor dimension table, modify fact table and link them together**

In [41]:
%%sql
DROP TABLE IF EXISTS Contributing_factor CASCADE;

CREATE TABLE Contributing_factor ( 
    key SERIAL PRIMARY KEY, 
    contributing_factor_vehicle_1 CHAR(100)
);

 * postgresql://student@/Final_Project
Done.
Done.


[]

Next is to create the second dimensional table "Contributing_factor", this dimensional table will have one primary key. 

CASCADE is often used with foreign keys and update delete commands. Meaning that the changes to the referenced table will automatically be applied to the referencing table. In other words this creates a parent-child relationship. 

In [42]:
%%sql
INSERT INTO Contributing_factor (contributing_factor_vehicle_1) 
SELECT DISTINCT contributing_factor_vehicle_1
FROM Crashes;

 * postgresql://student@/Final_Project
61 rows affected.


[]

Again, we will need to insert the contributing factors from the parent table "Crashes" to the child table 'Contributing_factors'. We'll then want to validate whether or not the insertion of data was done correctly. 

In [44]:
%%sql
SELECT * From Contributing_factor
Order By Key
limit 10

 * postgresql://student@/Final_Project
10 rows affected.


[(1, 'Animals Action                                                                                      '),
 (2, 'Steering Failure                                                                                    '),
 (3, 'Driver Inattention/Distraction                                                                      '),
 (4, 'Accelerator Defective                                                                               '),
 (5, 'Tire Failure/Inadequate                                                                             '),
 (6, 'Brakes Defective                                                                                    '),
 (7, 'Traffic Control Device Improper/Non-Working                                                         '),
 (8, 'Pedestrian/Bicyclist/Other Pedestrian Error/Confusion                                               '),
 (9, 'Illness                                                                                             '),
 (10, 'Failure to Keep Right                                                                               ')]

Next, step is to create the constraint key. 

In [45]:
%%sql
ALTER TABLE Crashes
ADD COLUMN contributing_factor_key INTEGER, 
ADD CONSTRAINT fk_contributing_factor
FOREIGN KEY (contributing_factor_key) REFERENCES Contributing_factor (key);

 * postgresql://student@/Final_Project
Done.


[]

In [46]:
%%sql
SELECT * from Crashes
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               ', 167, None),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               ', 217, None),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               ', 201, None),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               ', 8, None),
 (datetime.date(2021, 12, 14), datetime.time(16, 50), 'QUEENS              ', Decimal('11413'), Decimal('0'), Decimal('0'), 'Turning Improperly                                                                                  ', Decimal('4487127'), 'Sedan                                                                                               ', 69, None),
 (datetime.date(2021, 12, 14), datetime.time(23, 10), 'QUEENS              ', Decimal('11434'), Decimal('2'), Decimal('0'), 'Reaction to Uninvolved Vehicle                                                                      ', Decimal('4486635'), 'Sedan                                                                                               ', 59, None),
 (datetime.date(2021, 12, 14), datetime.time(12, 54), 'BROOKLYN            ', Decimal('11217'), Decimal('1'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4487052'), 'Sedan                                                                                               ', 174, None),
 (datetime.date(2021, 12, 14), datetime.time(17, 15), 'BROOKLYN            ', Decimal('11211'), Decimal('1'), Decimal('0'), 'Passing or Lane Usage Improper                                                                      ', Decimal('4486556'), 'Bus                                                                                                 ', 145, None),
 (datetime.date(2022, 3, 26), datetime.time(16, 2), 'QUEENS              ', Decimal('11373'), Decimal('1'), Decimal('0'), 'Lost Consciousness                                                                                  ', Decimal('4513794'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 220, None),
 (datetime.date(2022, 3, 26), datetime.time(16, 5), 'QUEENS              ', Decimal('11411'), Decimal('1'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4514267'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 98, None)]

In [47]:
%%sql
SELECT * from Contributing_factor
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(1, 'Animals Action                                                                                      '),
 (2, 'Steering Failure                                                                                    '),
 (3, 'Driver Inattention/Distraction                                                                      '),
 (4, 'Accelerator Defective                                                                               '),
 (5, 'Tire Failure/Inadequate                                                                             '),
 (6, 'Brakes Defective                                                                                    '),
 (7, 'Traffic Control Device Improper/Non-Working                                                         '),
 (8, 'Pedestrian/Bicyclist/Other Pedestrian Error/Confusion                                               '),
 (9, 'Illness                                                                                             '),
 (10, 'Failure to Keep Right                                                                               ')]

In [48]:
%%sql
UPDATE Crashes
SET contributing_factor_key = Contributing_factor.key
FROM Contributing_factor
WHERE 
    (Crashes.contributing_factor_vehicle_1 = Contributing_factor.contributing_factor_vehicle_1 OR
     (Crashes.contributing_factor_vehicle_1 IS NULL AND Contributing_factor.contributing_factor_vehicle_1 IS NULL))

 * postgresql://student@/Final_Project
793560 rows affected.


[]

This will insert the contraint key 'Contributing_factor.key' into the 'Crashes' table. This allows the exact values to be copied over. 

In [49]:
%%sql
SELECT count(*)
FROM Crashes
WHERE contributing_factor_key IS NULL

 * postgresql://student@/Final_Project
1 rows affected.


[(0,)]

Check there are no null values in the 'contributing_factor_key' column.

In [50]:
%%sql
SELECT count(DISTINCT contributing_factor_key)
FROM Crashes

 * postgresql://student@/Final_Project
1 rows affected.


[(61,)]

In [51]:
%%sql
SELECT count(key)
FROM Contributing_factor

 * postgresql://student@/Final_Project
1 rows affected.


[(61,)]

Check the distinct 'contributing_factor_key' count in both the fact table (Crashes) and the dimensional table (Contributing_factor). They have the same count, which is 61. Thus, we ensure that the fact table contains the right number of records.

**Work on Vehicle_type dimension table, modify fact table and link them together**

In [52]:
%%sql
DROP TABLE IF EXISTS Vehicle_type CASCADE;

CREATE TABLE Vehicle_type ( 
    key SERIAL PRIMARY KEY, 
    vehicle_type_code_1 CHAR(100)
);

 * postgresql://student@/Final_Project
Done.
Done.


[]

Now, we create the third dimensional table 'Vehicle_type': this dimensional table will have one primary key. 

CASCADE is often used with foreign keys and to update delete commands. Meaning if the changes to the referenced table will automatically be applied to the referencing table. In other words this creates a parent-child relationship.

In [53]:
%%sql
INSERT INTO Vehicle_type (vehicle_type_code_1) 
SELECT DISTINCT vehicle_type_code_1
FROM Crashes;

 * postgresql://student@/Final_Project
1145 rows affected.


[]

Again, we will need to insert the vehicle type from the parent table 'Crashes' to the child table 'Vehicle_type'. We'll then want to validate whether or not the insertion of data was done correctly.

In [54]:
%%sql
SELECT * From Vehicle_type
Order By Key
limit 10

 * postgresql://student@/Final_Project
10 rows affected.


[(1, 'SANTI                                                                                               '),
 (2, 'Tow t                                                                                               '),
 (3, 'commercial                                                                                          '),
 (4, 'HORSE CARR                                                                                          '),
 (5, 'FDNY Ambul                                                                                          '),
 (6, 'dumps                                                                                               '),
 (7, 'Sprinter W                                                                                          '),
 (8, 'jeep                                                                                                '),
 (9, 'Power shov                                                                                          '),
 (10, 'CONTR                                                                                               ')]

In [55]:
%%sql
ALTER TABLE Crashes
ADD COLUMN vehicle_type_key INTEGER, 
ADD CONSTRAINT fk_vehicle_type
FOREIGN KEY (vehicle_type_key) REFERENCES Vehicle_type (key);

 * postgresql://student@/Final_Project
Done.


[]

In [56]:
%%sql
SELECT * from Crashes
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               ', 167, 38, None),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               ', 217, 38, None),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               ', 201, 15, None),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               ', 8, 16, None),
 (datetime.date(2021, 12, 14), datetime.time(16, 50), 'QUEENS              ', Decimal('11413'), Decimal('0'), Decimal('0'), 'Turning Improperly                                                                                  ', Decimal('4487127'), 'Sedan                                                                                               ', 69, 34, None),
 (datetime.date(2021, 12, 14), datetime.time(23, 10), 'QUEENS              ', Decimal('11434'), Decimal('2'), Decimal('0'), 'Reaction to Uninvolved Vehicle                                                                      ', Decimal('4486635'), 'Sedan                                                                                               ', 59, 56, None),
 (datetime.date(2021, 12, 14), datetime.time(12, 54), 'BROOKLYN            ', Decimal('11217'), Decimal('1'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4487052'), 'Sedan                                                                                               ', 174, 38, None),
 (datetime.date(2021, 12, 14), datetime.time(17, 15), 'BROOKLYN            ', Decimal('11211'), Decimal('1'), Decimal('0'), 'Passing or Lane Usage Improper                                                                      ', Decimal('4486556'), 'Bus                                                                                                 ', 145, 21, None),
 (datetime.date(2022, 3, 26), datetime.time(16, 2), 'QUEENS              ', Decimal('11373'), Decimal('1'), Decimal('0'), 'Lost Consciousness                                                                                  ', Decimal('4513794'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 220, 46, None),
 (datetime.date(2022, 3, 26), datetime.time(16, 5), 'QUEENS              ', Decimal('11411'), Decimal('1'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4514267'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 98, 32, None)]

In [57]:
%%sql
SELECT * from Vehicle_type
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(1, 'SANTI                                                                                               '),
 (2, 'Tow t                                                                                               '),
 (3, 'commercial                                                                                          '),
 (4, 'HORSE CARR                                                                                          '),
 (5, 'FDNY Ambul                                                                                          '),
 (6, 'dumps                                                                                               '),
 (7, 'Sprinter W                                                                                          '),
 (8, 'jeep                                                                                                '),
 (9, 'Power shov                                                                                          '),
 (10, 'CONTR                                                                                               ')]

Note: This code will take around 3-5 min to run.

In [58]:
%%sql 
UPDATE Crashes
SET vehicle_type_key = Vehicle_type.key
FROM Vehicle_type
WHERE 
    (Crashes.vehicle_type_code_1 = Vehicle_type.vehicle_type_code_1 OR
     (Crashes.vehicle_type_code_1 IS NULL AND Vehicle_type.vehicle_type_code_1 IS NULL));

 * postgresql://student@/Final_Project
793560 rows affected.


[]

This will insert the contraint key 'Vehicle_type.key' into the 'Crashes' table. This allows the exact vales to be copied over.

In [59]:
%%sql
SELECT * FROM Crashes 
LIMIT 10;

 * postgresql://student@/Final_Project
10 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               ', 167, 38, 1108),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               ', 217, 38, 1108),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               ', 201, 15, 1108),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               ', 8, 16, 1108),
 (datetime.date(2021, 12, 14), datetime.time(16, 50), 'QUEENS              ', Decimal('11413'), Decimal('0'), Decimal('0'), 'Turning Improperly                                                                                  ', Decimal('4487127'), 'Sedan                                                                                               ', 69, 34, 1108),
 (datetime.date(2021, 12, 14), datetime.time(23, 10), 'QUEENS              ', Decimal('11434'), Decimal('2'), Decimal('0'), 'Reaction to Uninvolved Vehicle                                                                      ', Decimal('4486635'), 'Sedan                                                                                               ', 59, 56, 1108),
 (datetime.date(2021, 12, 14), datetime.time(12, 54), 'BROOKLYN            ', Decimal('11217'), Decimal('1'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4487052'), 'Sedan                                                                                               ', 174, 38, 1108),
 (datetime.date(2021, 12, 14), datetime.time(17, 15), 'BROOKLYN            ', Decimal('11211'), Decimal('1'), Decimal('0'), 'Passing or Lane Usage Improper                                                                      ', Decimal('4486556'), 'Bus                                                                                                 ', 145, 21, 1007),
 (datetime.date(2022, 3, 26), datetime.time(16, 2), 'QUEENS              ', Decimal('11373'), Decimal('1'), Decimal('0'), 'Lost Consciousness                                                                                  ', Decimal('4513794'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 220, 46, 1140),
 (datetime.date(2022, 3, 26), datetime.time(16, 5), 'QUEENS              ', Decimal('11411'), Decimal('1'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4514267'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 98, 32, 1140)]

In [60]:
%%sql
SELECT count(*) FROM Crashes 
Where vehicle_type_key is null

 * postgresql://student@/Final_Project
1 rows affected.


[(0,)]

Check there are no null values in the 'vehicle_type_key' column.

In [61]:
%%sql
SELECT count(DISTINCT vehicle_type_key)
FROM Crashes

 * postgresql://student@/Final_Project
1 rows affected.


[(1145,)]

In [62]:
%%sql
SELECT count(key)
FROM Vehicle_type

 * postgresql://student@/Final_Project
1 rows affected.


[(1145,)]

Check the distinct 'vehicle_type_key' count in both the fact table (Crashes) and the dimensional table (Vehicle_type). They have the same count, which is 1145. Thus, we ensure that the fact table contains the right number of records.

**Work on Crash_time dimension table, modify fact table and link them together**

In [63]:
%%sql
DROP TABLE IF EXISTS Crash_time CASCADE;

CREATE TABLE Crash_time ( 
    key SERIAL PRIMARY KEY, 
    hour INT,
    day INT,
    year INT,
    month_of_year INT, 
    day_of_month INT, 
    day_of_week_str CHAR(9), 
    day_of_week INT, 
    is_weekend BOOLEAN, 
    is_weekday BOOLEAN, 
    hour_of_day INT, 
    quarter_of_year INT,
    combined_timestamp TIMESTAMP
);

 * postgresql://student@/Final_Project
Done.
Done.


[]

Finally, we create of last dimensional table 'Crash_time'. This table is important to our queries, this will house our weeks, and time variables.

In [64]:
%%sql
INSERT INTO Crash_time (hour, day, year, month_of_year, day_of_month, day_of_week_str, day_of_week, is_weekend, is_weekday, hour_of_day, quarter_of_year, combined_timestamp) 
SELECT DISTINCT 
    EXTRACT(HOUR FROM combined_timestamp) AS hour, 
    EXTRACT(DAY FROM combined_timestamp) AS day,
    EXTRACT(YEAR FROM combined_timestamp) AS year,
    EXTRACT(MONTH FROM combined_timestamp) AS month_of_year,
    EXTRACT(DAY FROM combined_timestamp) AS day_of_month,
    TO_CHAR(combined_timestamp, 'Day') AS day_of_week_str,
    EXTRACT(DOW FROM combined_timestamp) AS day_of_week,
    CASE WHEN EXTRACT(DOW FROM combined_timestamp) IN (0, 6) THEN TRUE ELSE FALSE END AS is_weekend,
    CASE WHEN EXTRACT(DOW FROM combined_timestamp) NOT IN (0,6) THEN TRUE ELSE FALSE END AS is_weekday,
    EXTRACT(HOUR FROM combined_timestamp) AS hour_of_day,
    CEIL(EXTRACT(MONTH FROM combined_timestamp) / 3.0) AS quarter_of_year,
    combined_timestamp
FROM 
    (SELECT crash_date + crash_time AS combined_timestamp FROM Crashes) AS combined;

 * postgresql://student@/Final_Project
501871 rows affected.


[]

Here we are inserting the data into this table. Then, 'weekdays' to a specific numeric value and sequence so weekdays will have number 1,2,3,4,5 and weekends as 0,6. 

In [65]:
%%sql
SELECT DISTINCT day_of_week_str, day_of_week, is_weekend, is_weekday 
FROM Crash_time
ORDER BY day_of_week;

 * postgresql://student@/Final_Project
7 rows affected.


[('Sunday   ', 0, True, False),
 ('Monday   ', 1, False, True),
 ('Tuesday  ', 2, False, True),
 ('Wednesday', 3, False, True),
 ('Thursday ', 4, False, True),
 ('Friday   ', 5, False, True),
 ('Saturday ', 6, True, False)]

This validates that our weekdays and weekends are being read correctly.

In [66]:
%%sql
ALTER TABLE Crashes
ADD COLUMN crash_time_key INTEGER, 
ADD CONSTRAINT fk_crash_time
FOREIGN KEY (crash_time_key) REFERENCES Crash_time (key);

 * postgresql://student@/Final_Project
Done.


[]

We add 'crash_time_key' to the 'Crashes' table and add the constraint. This allows us to run the below query which shows us the values 'crash_time_key' contains. 

In [67]:
%%sql
SELECT * From Crashes
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(datetime.date(2021, 9, 11), datetime.time(9, 35), 'BROOKLYN            ', Decimal('11208'), Decimal('0'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4456314'), 'Sedan                                                                                               ', 167, 38, 1108, None),
 (datetime.date(2021, 12, 14), datetime.time(8, 17), 'BRONX               ', Decimal('10475'), Decimal('2'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4486660'), 'Sedan                                                                                               ', 217, 38, 1108, None),
 (datetime.date(2021, 12, 14), datetime.time(21, 10), 'BROOKLYN            ', Decimal('11207'), Decimal('0'), Decimal('0'), 'Driver Inexperience                                                                                 ', Decimal('4487074'), 'Sedan                                                                                               ', 201, 15, 1108, None),
 (datetime.date(2021, 12, 14), datetime.time(14, 58), 'MANHATTAN           ', Decimal('10017'), Decimal('0'), Decimal('0'), 'Passing Too Closely                                                                                 ', Decimal('4486519'), 'Sedan                                                                                               ', 8, 16, 1108, None),
 (datetime.date(2021, 12, 14), datetime.time(16, 50), 'QUEENS              ', Decimal('11413'), Decimal('0'), Decimal('0'), 'Turning Improperly                                                                                  ', Decimal('4487127'), 'Sedan                                                                                               ', 69, 34, 1108, None),
 (datetime.date(2021, 12, 14), datetime.time(23, 10), 'QUEENS              ', Decimal('11434'), Decimal('2'), Decimal('0'), 'Reaction to Uninvolved Vehicle                                                                      ', Decimal('4486635'), 'Sedan                                                                                               ', 59, 56, 1108, None),
 (datetime.date(2021, 12, 14), datetime.time(12, 54), 'BROOKLYN            ', Decimal('11217'), Decimal('1'), Decimal('0'), 'Unspecified                                                                                         ', Decimal('4487052'), 'Sedan                                                                                               ', 174, 38, 1108, None),
 (datetime.date(2021, 12, 14), datetime.time(17, 15), 'BROOKLYN            ', Decimal('11211'), Decimal('1'), Decimal('0'), 'Passing or Lane Usage Improper                                                                      ', Decimal('4486556'), 'Bus                                                                                                 ', 145, 21, 1007, None),
 (datetime.date(2022, 3, 26), datetime.time(16, 2), 'QUEENS              ', Decimal('11373'), Decimal('1'), Decimal('0'), 'Lost Consciousness                                                                                  ', Decimal('4513794'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 220, 46, 1140, None),
 (datetime.date(2022, 3, 26), datetime.time(16, 5), 'QUEENS              ', Decimal('11411'), Decimal('1'), Decimal('0'), 'Following Too Closely                                                                               ', Decimal('4514267'), 'Station Wagon/Sport Utility Vehicle                                                                 ', 98, 32, 1140, None)]

In [68]:
%%sql
SELECT * from Crash_time
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(1, 0, 1, 2016, 1, 1, 'Friday   ', 5, False, True, 0, 1, datetime.datetime(2016, 1, 1, 0, 1)),
 (2, 0, 1, 2016, 1, 1, 'Friday   ', 5, False, True, 0, 1, datetime.datetime(2016, 1, 1, 0, 5)),
 (3, 0, 1, 2016, 1, 1, 'Friday   ', 5, False, True, 0, 1, datetime.datetime(2016, 1, 1, 0, 15)),
 (4, 0, 1, 2016, 1, 1, 'Friday   ', 5, False, True, 0, 1, datetime.datetime(2016, 1, 1, 0, 23)),
 (5, 0, 1, 2016, 1, 1, 'Friday   ', 5, False, True, 0, 1, datetime.datetime(2016, 1, 1, 0, 25)),
 (6, 0, 1, 2016, 1, 1, 'Friday   ', 5, False, True, 0, 1, datetime.datetime(2016, 1, 1, 0, 33)),
 (7, 0, 1, 2016, 1, 1, 'Friday   ', 5, False, True, 0, 1, datetime.datetime(2016, 1, 1, 0, 50)),
 (8, 0, 1, 2016, 1, 1, 'Friday   ', 5, False, True, 0, 1, datetime.datetime(2016, 1, 1, 0, 51)),
 (9, 0, 1, 2016, 2, 1, 'Monday   ', 1, False, True, 0, 1, datetime.datetime(2016, 2, 1, 0, 1)),
 (10, 0, 1, 2016, 2, 1, 'Monday   ', 1, False, True, 0, 1, datetime.datetime(2016, 2, 1, 0, 5))]

In [69]:
%%sql
UPDATE Crashes
SET crash_time_key = Crash_time.key
FROM Crash_time
WHERE Crashes.crash_date + Crashes.crash_time = Crash_time.combined_timestamp;

 * postgresql://student@/Final_Project
793560 rows affected.


[]

This will insert the contraint key 'Crash_time.key' into the 'Crashes' table. This allows the exact vales to be copied over. 

We build another contraint key 'Crash_time.combined_timestamp'. This will combine both the date and time into one column in the 'Crashes' table.

In [70]:
%%sql
SELECT count(*) From Crashes
Where crash_time_key is NULL

 * postgresql://student@/Final_Project
1 rows affected.


[(0,)]

Check there are no null values in the 'crash_time_key' column.

In [71]:
%%sql
SELECT count(DISTINCT crash_time_key)
FROM Crashes

 * postgresql://student@/Final_Project
1 rows affected.


[(501871,)]

In [72]:
%%sql
SELECT count(key)
FROM Crash_time

 * postgresql://student@/Final_Project
1 rows affected.


[(501871,)]

Check the distinct 'crash_time_key' count in both the fact table (Crashes) and the dimensional table (Crash_time). They have the same count, which is 501871. Thus, we ensure that the fact table contains the right number of records.

In [73]:
%%sql
ALTER TABLE Crashes
DROP COLUMN crash_date,
DROP COLUMN crash_time,
DROP COLUMN borough,
DROP COLUMN zip_code,
DROP COLUMN contributing_factor_vehicle_1,
DROP COLUMN vehicle_type_code_1

 * postgresql://student@/Final_Project
Done.


[]

Drop the original and unnecessary columns in the fact table.

**Look at the final fact table.**

In [74]:
%%sql
SELECT * From Crashes
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(Decimal('0'), Decimal('0'), Decimal('3363638'), 89, 9, 826, 549),
 (Decimal('0'), Decimal('0'), Decimal('3364381'), 56, 38, 207, 142114),
 (Decimal('0'), Decimal('0'), Decimal('3364154'), 1, 38, 207, 61933),
 (Decimal('1'), Decimal('0'), Decimal('3364847'), 49, 61, 207, 118146),
 (Decimal('0'), Decimal('0'), Decimal('3364309'), 88, 38, 804, 308219),
 (Decimal('0'), Decimal('0'), Decimal('3364599'), 209, 54, 883, 221645),
 (Decimal('0'), Decimal('0'), Decimal('3365018'), 215, 61, 207, 341302),
 (Decimal('1'), Decimal('0'), Decimal('3365321'), 212, 57, 826, 449301),
 (Decimal('0'), Decimal('0'), Decimal('3592792'), 119, 3, 1108, 62810),
 (Decimal('0'), Decimal('0'), Decimal('3365546'), 200, 3, 826, 76390)]

### Ask 3: Data Analysis

We are a Consultation firm providing an overview of the New York City traffic incident trends from 2016-2022.

In [21]:
image_url = 'https://i.ibb.co/FHPM3BG/Number-Of-Incidents.jpg'
Image(url=image_url)

We can see that this is the total number of incidents from 2016- 2022. The lowest number of incidents is in 2020 (April) and we can see that the number of incidents has decreased significantly post 2020, which could possibly be because of post CoViD circumstances (e,g; more people choosing to Work from Home).

In [20]:
image_url = 'https://i.ibb.co/6wbprvL/Total-Persons-Injured.jpg'
Image(url=image_url)

- The trend shows that even though the total number of incidents has decreased, we can see that the trend in the number of injuries is not similar (decreased).  


In [22]:
image_url = 'https://i.ibb.co/p0LMkNC/Total-Persons-Killed.jpg'
Image(url=image_url)

- The trend shows that even though the total number of incidents has decreased, we can see that the trend in the number of people killed remains the same. 

The following queries are intended for data extraction to generate the overall trend graphs above.

In [75]:
!pwd

/home/ubuntu/notebooks/FinalProjectGroup2


In [ ]:
!sudo chmod a+w /home/ubuntu/notebooks/Final\ Project

In [ ]:
%%sql
COPY (
    SELECT
        Crash_time.year, 
        Crash_time.month_of_year,
        Crash_time.day_of_week_str,
        Geography.borough,
        count(Crashes.collision_id) AS number_of_incidents,
        SUM(Crashes.number_of_persons_killed) AS total_persons_killed,
        SUM(Crashes.number_of_persons_injured) AS total_persons_injured    
    FROM Crashes
    JOIN Crash_time ON Crashes.crash_time_key = Crash_time.key
    JOIN Geography ON Crashes.geography_key = Geography.key
    GROUP BY Crash_time.year, Crash_time.month_of_year, Crash_time.day_of_week_str, Geography.borough
    ORDER BY Crash_time.year, Crash_time.month_of_year, Crash_time.day_of_week_str, Geography.borough, number_of_incidents
) TO '/home/ubuntu/notebooks/output.csv' WITH CSV HEADER;

In [ ]:
%%sql
COPY (
    SELECT
    Contributing_factor.contributing_factor_vehicle_1,
    SUM(Crashes.number_of_persons_killed) AS total_persons_killed,
    SUM(Crashes.number_of_persons_injured) AS total_persons_injured
FROM
    Crashes
JOIN
    Contributing_factor ON Crashes.contributing_factor_key = Contributing_factor.key
GROUP BY
    Contributing_factor.contributing_factor_vehicle_1
ORDER BY
   total_persons_killed DESC, total_persons_injured DESC
) TO '/home/ubuntu/notebooks/output1.csv' WITH CSV HEADER;

#### Analysis 1 for Car Insurance:

Examine the boroughs with the highest frequency of  traffic crashes and the vehicles involved in the highest  number of accidents focusing on developing  recommendations for car insurance companies by improving risk assessment models and enhancing competitiveness within the insurance market.

**First we want to know which top 10 vehicle types have the most crashes**

In [76]:
%%sql
SELECT Vehicle_type.vehicle_type_code_1, count(collision_id) as number_of_incidents
FROM
    Crashes 
JOIN
    Vehicle_type ON Crashes.vehicle_type_key = Vehicle_type.key
    
GROUP BY vehicle_type.vehicle_type_code_1
ORDER BY number_of_incidents desc
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[('Sedan                                                                                               ', 335762),
 ('Station Wagon/Sport Utility Vehicle                                                                 ', 262104),
 ('Taxi                                                                                                ', 31858),
 ('4 dr sedan                                                                                          ', 26889),
 ('Pick-up Truck                                                                                       ', 20100),
 ('PASSENGER VEHICLE                                                                                   ', 15575),
 ('Box Truck                                                                                           ', 15067),
 ('Bus                                                                                                 ', 13226),
 ('Bike                                                                                                ', 8884),
 ('SPORT UTILITY / STATION WAGON                                                                       ', 8230)]

As it shows that there are some duplications in Vehicle type, due to the fact that no standards are followed by the people who did data entry. This results in the problem that each person enters the data differently, some will say Sedan, other will enter the car company name like 'chev' or 'ford'.

To solve this issue, we need first to identify and agree on the vehicle type classfication. Second, we need to automate the data entry so the data entry person will have a drop down list to chose the vehicle type.


 - Sedan vehicles top the list for the highest number of incidents, closely followed by Station Wagons and Sport Utility Vehicles.

In [23]:
image_url = 'https://i.ibb.co/3FwJ2L0/Vehicle-Types.jpg'
Image(url=image_url)

**Which boroughs have the most crashes?**

In [77]:
%%sql
SELECT Geography.borough, 
count(collision_id) as number_of_incidents
FROM
    Crashes 
JOIN
    Geography ON Crashes.geography_key = Geography.key
    
GROUP BY Geography.borough
ORDER BY number_of_incidents desc

 * postgresql://student@/Final_Project
5 rows affected.


[('BROOKLYN            ', 256981),
 ('QUEENS              ', 218871),
 ('MANHATTAN           ', 159491),
 ('BRONX               ', 127786),
 ('STATEN ISLAND       ', 30431)]

In [24]:
image_url = 'https://i.ibb.co/L68CZqY/Borough.jpg'
Image(url=image_url)

We find that 'BROOKLYN' and 'QUEENS' have the highest number of incidents

**Strategic recommendations to Car Insurance Companies:**

- We suggest raising insurance premiums for Sedans and SPORT UTILITY / STATION WAGON vehicle types, as well as for residents of BROOKLYN and QUEENS.

#### Analysis 2 for NY Traffic Police Office:

What recommendations can we provide to the New York Police Office responsible for traffic management, with the goal of reducing traffic accidents in high-risk areas?

**2-1: Which top 10 zip code areas have the highest number of persons injured due to crashes?**

In [78]:
%%sql
SELECT Geography.zip_code, 
SUM(Crashes.number_of_persons_injured) AS total_persons_injured
FROM
    Crashes 
JOIN
    Geography ON Crashes.geography_key = Geography.key
    
GROUP BY Geography.zip_code
ORDER BY total_persons_injured desc
LIMIT 10

 * postgresql://student@/Final_Project
10 rows affected.


[(Decimal('11207'), Decimal('6060')),
 (Decimal('11236'), Decimal('4998')),
 (Decimal('11203'), Decimal('4522')),
 (Decimal('11212'), Decimal('4380')),
 (Decimal('11234'), Decimal('3804')),
 (Decimal('11208'), Decimal('3787')),
 (Decimal('11226'), Decimal('3782')),
 (Decimal('11434'), Decimal('3545')),
 (Decimal('11233'), Decimal('3111')),
 (Decimal('11385'), Decimal('3049'))]

In [16]:
image_url = 'https://i.ibb.co/XLX9Kfj/Zip-Code-Map.jpg'

Image(url=image_url)

We use this website, https://maps.huge.info/zip.htm, to pinpoint the top 10 high-risk zip codes. The pinned locations are as follows: Pinned A: 11207, Pinned B: 11236, Pinned C: 11203, Pinned D: 11212, Pinned E: 11234, Pinned F: 11208, Pinned G: 11226, Pinned H: 11434, Pinned I: 11233, and Pinned J: 11385.

We recommend that the local police office initiate research to identify specific intersections within the identified high-risk zip codes (11207, 11236, 11203, 11212, 11234, 11208, 11226, 11434, 11208, 11226, 11434, 11233, 11385) and implement safety measures such as improved signage, road markings, and pedestrian crossings. Simultaneously, we suggest conducting traffic awareness campaigns in these areas to educate both drivers and pedestrians about safe practices.

**2-2: What combination of day of the week, hour of the day, and zip code experienced the top 30 highest number of persons injured in crashes?**

In [79]:
%%sql
SELECT DISTINCT
    Crash_time.day_of_week_str,
    Crash_time.hour_of_day,
    Geography.zip_code,
    SUM(Crashes.number_of_persons_injured) AS total_persons_injured
FROM
    Crashes
JOIN
    Crash_time ON Crashes.crash_time_key = Crash_time.key
JOIN
    Geography ON Crashes.geography_key = Geography.key
GROUP BY
    Crash_time.day_of_week_str, Crash_time.hour_of_day, Geography.zip_code
ORDER BY
    total_persons_injured DESC
limit 30

 * postgresql://student@/Final_Project
30 rows affected.


[('Friday   ', 17, Decimal('11207'), Decimal('96')),
 ('Sunday   ', 0, Decimal('11207'), Decimal('75')),
 ('Thursday ', 16, Decimal('11207'), Decimal('72')),
 ('Saturday ', 22, Decimal('11207'), Decimal('69')),
 ('Sunday   ', 15, Decimal('11207'), Decimal('68')),
 ('Tuesday  ', 17, Decimal('11234'), Decimal('66')),
 ('Tuesday  ', 18, Decimal('11207'), Decimal('66')),
 ('Saturday ', 21, Decimal('11236'), Decimal('65')),
 ('Thursday ', 17, Decimal('11207'), Decimal('65')),
 ('Saturday ', 15, Decimal('11207'), Decimal('64')),
 ('Saturday ', 18, Decimal('11236'), Decimal('63')),
 ('Thursday ', 17, Decimal('11434'), Decimal('63')),
 ('Friday   ', 14, Decimal('11207'), Decimal('62')),
 ('Friday   ', 16, Decimal('11207'), Decimal('61')),
 ('Friday   ', 20, Decimal('11207'), Decimal('61')),
 ('Friday   ', 20, Decimal('11236'), Decimal('61')),
 ('Sunday   ', 0, Decimal('11236'), Decimal('61')),
 ('Tuesday  ', 17, Decimal('10451'), Decimal('61')),
 ('Wednesday', 15, Decimal('11203'), Decimal('61')),
 ('Friday   ', 18, Decimal('11385'), Decimal('60')),
 ('Friday   ', 15, Decimal('11207'), Decimal('58')),
 ('Friday   ', 18, Decimal('11207'), Decimal('58')),
 ('Saturday ', 19, Decimal('11236'), Decimal('58')),
 ('Thursday ', 14, Decimal('11207'), Decimal('58')),
 ('Thursday ', 17, Decimal('11203'), Decimal('58')),
 ('Thursday ', 17, Decimal('11234'), Decimal('58')),
 ('Thursday ', 18, Decimal('11212'), Decimal('58')),
 ('Tuesday  ', 17, Decimal('11208'), Decimal('58')),
 ('Tuesday  ', 18, Decimal('11203'), Decimal('58')),
 ('Tuesday  ', 20, Decimal('11207'), Decimal('58'))]

**Police Traffic Command Schedule**

| Time        | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday | Sunday |
|-------------|--------|---------|-----------|----------|--------|----------|--------|
| 0:00-1:00   |        |         |           |          |        |          | 11207 & 11236  |
| 1:00-2:00   |        |         |           |          |        |          |        |
| 2:00-3:00   |        |         |           |          |        |          |        |
| 3:00-4:00   |        |         |           |          |        |          |        |
| 4:00-5:00   |        |         |           |          |        |          |        |
| 5:00-6:00   |        |         |           |          |        |          |        |
| 6:00-7:00   |        |         |           |          |        |          |        |
| 7:00-8:00   |        |         |           |          |        |          |        |
| 8:00-9:00   |        |         |           |          |        |          |        |
| 9:00-10:00  |        |         |           |          |        |          |        |
| 10:00-11:00 |        |         |           |          |        |          |        |
| 11:00-12:00 |        |         |           |          |        |          |        |
| 12:00-13:00 |        |         |           |          |        |          |        |
| 13:00-14:00 |        |         |           |          |        |          |        |
| 14:00-15:00 |        |         |           | 11207          | 11207  |          |        |
| 15:00-16:00 |        |         | 11203     |          | 11207        | 11207    | 11207  |
| 16:00-17:00 |        |         |           | 11207    | 11207  |          |        |
| 17:00-18:00 |        | 11234 & 10451 & 11208   |     11234 & 11203      | 11207 & 11434    | 11207  |          |        |
| 18:00-19:00 |        | 11207 & 11203|           | 11212          | 11385 & 11207  | 11236    |        |
| 19:00-20:00 |        |         |           |          |        | 11236    |        |
| 20:00-21:00 |        | 11207         |           |          | 11207 & 11236        |          |        |
| 21:00-22:00 |        |         |           |          |        | 11236    |        |
| 22:00-23:00 |        |         |           |          |        | 11207    |        |
| 23:00-0:00  |        |         |           |          |        |          |        |

We recommend that the local police office **allocate additional police presence** to the identified combinations of the day of the week, hour of the day, and zip code with high-risk crash incidents based on the above schedule. For example, the analysis indicates that Friday at 5:00 PM in the zip code 11207 has the highest number of persons injured due to car crashes. The police department could consider increasing patrols and visibility during this specific time and in this particular zip code.

Furthermore, we suggest **implement more security cameras to monitor and track vehicle speeds and behaviour** would enhance overall road safety.

#### Analysis 3 for New York Department of Motor Vehicles (DMV):

**Identify the top 10 contributing factors to the highest number of fatalities** and leverage this critical information to provide strategic recommendations to NY Department of Motor Vehicles (DMV). Enable the implementation of targeted interventions and policies aimed at addressing the identified factors and reducing the overall impact of road incidents on public safety.

In [80]:
%%sql
SELECT
    Contributing_factor.contributing_factor_vehicle_1,
    SUM(Crashes.number_of_persons_killed) AS total_persons_killed,
    SUM(Crashes.number_of_persons_injured) AS total_persons_injured
FROM
    Crashes
JOIN
    Contributing_factor ON Crashes.contributing_factor_key = Contributing_factor.key
GROUP BY
    Contributing_factor.contributing_factor_vehicle_1
ORDER BY
   total_persons_killed DESC, total_persons_injured DESC
LIMIT 10;


 * postgresql://student@/Final_Project
10 rows affected.


[('Unspecified                                                                                         ', Decimal('259'), Decimal('51456')),
 ('Failure to Yield Right-of-Way                                                                       ', Decimal('150'), Decimal('31632')),
 ('Driver Inattention/Distraction                                                                      ', Decimal('125'), Decimal('59476')),
 ('Unsafe Speed                                                                                        ', Decimal('123'), Decimal('7731')),
 ('Traffic Control Disregarded                                                                         ', Decimal('78'), Decimal('11997')),
 ('Pedestrian/Bicyclist/Other Pedestrian Error/Confusion                                               ', Decimal('45'), Decimal('4538')),
 ('Alcohol Involvement                                                                                 ', Decimal('25'), Decimal('4448')),
 ('Driver Inexperience                                                                                 ', Decimal('24'), Decimal('3963')),
 ('Illnes                                                                                              ', Decimal('19'), Decimal('780')),
 ('Backing Unsafely                                                                                    ', Decimal('18'), Decimal('4810'))]

**Analysis and Recommendations for Top 10 factors:**

1. Unspecified:

Recommendation: Improve data recording practices to specify contributing factors accurately. Enhance awareness campaigns to encourage reporting details of incidents in an automated way.

2. Failure to Yield Right-of-Way:

Recommendation: Strengthen education and enforcement regarding right-of-way rules. Conduct targeted driver education programs emphasizing yielding protocols.

3. Driver Inattention/Distraction:

Recommendation: Implement stricter penalties for distracted driving. Promote public awareness on the dangers of distracted driving and encourage the use of hands-free technology.

4. Unsafe Speed:

Recommendation: Increase enforcement of speed limits. Implement speed-calming measures in high-risk areas. Launch campaigns emphasizing the correlation between speed and accident severity.

5. Traffic Control Disregarded:

Recommendation: Enhance traffic control infrastructure and visibility. Enforce strict penalties for violations. Conduct public awareness campaigns on the importance of obeying traffic signals.

6. Pedestrian/Bicyclist/Other Pedestrian Error/Confusion:

Recommendation: Invest in pedestrian and cyclist safety infrastructure. Develop educational programs for both drivers and pedestrians to reduce confusion and errors.

7. Alcohol Involvement:

Recommendation: Intensify DUI enforcement. Increase public awareness on the dangers of drinking and driving. Implement stricter penalties for alcohol-related offenses.

8. Driver Inexperience:

Recommendation: Strengthen driver education programs, especially for new drivers. Consider implementing graduated licensing systems.

9. Illness:

Recommendation: Encourage individuals with known medical conditions to assess their fitness to drive regularly. Raise awareness among healthcare professionals regarding reporting concerns about patients' ability to drive.

10. Backing Unsafely:

Recommendation: Promote safe backing practices through educational campaigns. Consider technological solutions, such as backup cameras, to assist drivers in preventing accidents.

**Overall Recommendations:**

a) Data Quality Improvement: Enhance data collection methods to ensure accurate and detailed recording of contributing factors.

b) Education and Awareness: Launch targeted public awareness campaigns addressing specific contributing factors.

c) Enforcement: Strengthen law enforcement measures, especially for high-impact factors like distracted driving and speeding.

d) Infrastructure Improvement: Invest in infrastructure enhancements, such as improved traffic control and pedestrian safety measures.

e) Policy Development: Consider revising and implementing policies that address the identified contributing factors effectively.

f) Implementing a multi-faceted approach, combining education, enforcement, and infrastructure improvements, will likely yield the most significant impact in reducing fatalities and injuries related to road incidents.


**Strategic recommendations to NY DMV:**

- As per current Licence Violation Points, the point deduction for 'Failure to Yield Right-of-Way'is 3. We recommended an increase of point-deduction to 5 to implement stricter regulations. Based on our analysis, we see that the total number of people killed by this contributing factor for the last 7 years is 150.

- Based on our analysis, we see that the total number of people killed by 'Driver Inattention/Distraction' is more than those killed by 'Unsafe Speed'. As per current Licence Violation Points, the point deduction for 'Improper cellphone usage' & 'Use of portable electronic device 'Texting' is 5 points whereas, for 'Speeding', it is 3-11 points. We recommend the DMV to increase the points deducted for this particular factor.

##### Below is the screenshot from the NY DMV Official website (https://dmv.ny.gov/tickets/about-nys-driver-point-system) for License Violation Pionts Deduction:

In [81]:
image_url = 'https://i.ibb.co/w7wpqJp/Whats-App-Image-2023-12-09-at-12-24-16-PM.jpg'
Image(url=image_url)

In [54]:
!sudo chmod a+w /home/ubuntu/notebooks

In [82]:
!pwd

/home/ubuntu/notebooks/FinalProjectGroup2


In [55]:
%%sql
COPY (
    SELECT
        Crash_time.year, 
        Crash_time.month_of_year,
        Crash_time.day_of_week_str,
        Geography.borough,
        count(Crashes.collision_id) AS number_of_incidents,
        SUM(Crashes.number_of_persons_killed) AS total_persons_killed,
        SUM(Crashes.number_of_persons_injured) AS total_persons_injured    
    FROM Crashes
    JOIN Crash_time ON Crashes.crash_time_key = Crash_time.key
    JOIN Geography ON Crashes.geography_key = Geography.key
    GROUP BY Crash_time.year, Crash_time.month_of_year, Crash_time.day_of_week_str, Geography.borough
    ORDER BY Crash_time.year, Crash_time.month_of_year, Crash_time.day_of_week_str, Geography.borough, number_of_incidents
) TO '/home/ubuntu/notebooks/output.csv' WITH CSV HEADER;

 * postgresql://student@/Final_Project
2940 rows affected.


[]

In [23]:
%%sql
COPY (
    SELECT
    Contributing_factor.contributing_factor_vehicle_1,
    SUM(Crashes.number_of_persons_killed) AS total_persons_killed,
    SUM(Crashes.number_of_persons_injured) AS total_persons_injured
FROM
    Crashes
JOIN
    Contributing_factor ON Crashes.contributing_factor_key = Contributing_factor.key
GROUP BY
    Contributing_factor.contributing_factor_vehicle_1
ORDER BY
   total_persons_killed DESC, total_persons_injured DESC
) TO '/home/ubuntu/notebooks/output1.csv' WITH CSV HEADER;

 * postgresql://student@/Final_Project
61 rows affected.


[]